src/iter/parallel.rs

```diff
@@ -44,56 +43,61 @@ impl<'a, T: 'a + Send + Sync> UnindexedProducer for BitProducer<'a, T>
-                    self.0.masks[level] = 0;
-                    // Get the mask of the child layer from the set
-                    self.0.masks[level - 1] = get_from_layer(self.0.set, level - 1, idx);
-                    return None;
-                }
-                // Make the split point to be the average of first and last bit.
-                let after_average = (first_bit + last_bit) / 2 + 1;
-                // A bit mask to get the lower half of the mask
-                // This cuts the mask to half from the average
-                let mask = (1 << after_average) - 1;
-                let mut other = BitProducer(BitIter::new(self.0.set, [0, 0, 0, 0], [0, 0, 0]));
-                let original_mask = self.0.masks[level];
-                // Take the higher half of the mask
-                other.0.masks[level] = original_mask & !mask;
-                // The higher levels of masks don't need to preserved as they are empty.
-                for n in &self.0.masks[(level + 1)..] {
-                    debug_assert_eq!(*n, 0);
-                }
-                // The higher half starts iterating after the average
-                other.0.prefix[level - 1] = (level_prefix | after_average) << BITS;
-                // And preserve the prefix of the higher levels
-                other.0.prefix[level..].copy_from_slice(&self.0.prefix[level..]);
-                // Take the lower half the mask
-                self.0.masks[level] = original_mask & mask;
-                // Combined mask of the current level should now equal the original mask
-                debug_assert_eq!(self.0.masks[level] | other.0.masks[level], original_mask);
-                // The lower half starts iterating from the first bit
-                self.0.prefix[level - 1] = (level_prefix | first_bit) << BITS;
-                Some(other)
+                let first_bit = self.0.masks[level].trailing_zeros();
+                average_ones(self.0.masks[level])
+                    .and_then(|average_bit| {
+                        let mask = (1 << (average_bit - 1)) - 1;
+                        let mut other = BitProducer(BitIter::new(self.0.set, [0; 4], [0; 3]));
+                        // `other` is the more significant half of the mask
+                        other.0.masks[level] = self.0.masks[level] & !mask;
+                        other.0.prefix[level - 1] = (level_prefix | average_bit as u32) << BITS;
+                        other.0.prefix[level..].copy_from_slice(&self.0.prefix[level..]);
+                        // And `self` is the less significant one
+                        self.0.masks[level] &= mask;
+                        self.0.prefix[level - 1] = (level_prefix | first_bit) << BITS;
+                        Some(other)
+                    }).or_else(|| {
+                        // Because there is only one bit left we descend to it
+                        let idx = level_prefix as usize | first_bit as usize;
```

Comment:
this logic just becomes one giant magic blob to me, tbh
We need more in-code documentation diagrams and explanations to make this maintenable

Reply:
What exactly do you think is magic?

Reply:
What is the hardest thing to understand from it? I have been reading/writing the code, so I am not sure what to improve.

Reply:
it's just a solid piece of bit-mangling code
my old brain requires extra focus/energy to go through it :/